### PR TITLE
[SPARK-16311][SQL] Metadata refresh should work on temporary views

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

This patch fixes the bug that the refresh command does not work on temporary views. This patch is based on https://github.com/apache/spark/pull/13989, but removes the public Dataset.refresh() API as well as improved test coverage.

Note that I actually think the public refresh() API is very useful. We can in the future implement it by also invalidating the lazy vals in QueryExecution (or alternatively just create a new QueryExecution).
## How was this patch tested?

Re-enabled a previously ignored test, and added a new test suite for Hive testing behavior of temporary views against MetastoreRelation.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -462,17 +462,15 @@ class SessionCatalog(
...
    * Refresh the cache entry for a metastore table, if any.
    */
-  def refreshTable(name: TableIdentifier): Unit = { /* no-op */ }
-
-  /**
-   * Invalidate the cache entry for a metastore table, if any.
-   */
-  def invalidateTable(name: TableIdentifier): Unit = { /* no-op */ }
+  def refreshTable(name: TableIdentifier): Unit = {
+    // Go through temporary tables and invalidate them.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: In the test case of `HiveMetadataCacheSuite.scala`, users might refresh the base table by using `spark.catalog.refreshTable("view_refresh")`. Normally, they do not specify the current database name. Then, its database name is empty. Thus, this table will be treated as a temporary table. This comment might need a correction.

**Comment by gatorsmile** [comment]
LGTM except a minor comment.

File: sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala [hunk 2]
```diff
@@ -77,12 +77,12 @@ class MetadataCacheSuite extends QueryTest with SharedSQLContext {
...
       assert(e.getMessage.contains("FileNotFoundException"))
-      assert(e.getMessage.contains("refresh()"))
+      assert(e.getMessage.contains("REFRESH"))
 
       // Refresh and we should be able to read it again.
       spark.catalog.refreshTable("view_refresh")
       val newCount = sql("select count(*) from view_refresh").first().getLong(0)
       assert(newCount > 0 && newCount < 100)
-    }
+    }}
```
Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/MetadataCacheSuite.scala]: This style is pretty weird...

**Comment by liancheng** [comment]
LGTM except for a minor styling issue.

**Comment by rxin** [comment]
Thanks - I fixed the two comments. Going to merge it in master/2.0.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
It seems this causes build failure.
https://amplab.cs.berkeley.edu/jenkins/job/spark-master-compile-maven-hadoop-2.7/2094/console

**Comment by dongjoon-hyun** [comment]
Thank you for fast fix!

**Pull request closed without merging**